### DISCUSSION TITLE
Why is indent rule so unbearably slow?

### DISCUSSION THREAD

**Discussion by dumbasPL** [eslint/eslint#16032] [Q&A] [answered]
I've noticed that eslint has been really slowing down for me lately so i decided to investigate and found the `TIMING=1` env variable. Here are the results:

```shell
$ TIMING=1 npx eslint ./
Rule                          | Time (ms) | Relative
:-----------------------------|----------:|--------:
indent                        |   656.331 |    76.6%
no-unused-vars                |    28.158 |     3.3%
no-unexpected-multiline       |    13.333 |     1.6%
no-regex-spaces               |    12.633 |     1.5%
no-dupe-keys                  |    12.343 |     1.4%
no-loss-of-precision          |    10.243 |     1.2%
no-redeclare                  |     9.984 |     1.2%
no-misleading-character-class |     7.835 |     0.9%
no-unsafe-optional-chaining   |     7.640 |     0.9%
semi                          |     7.529 |     0.9%
```

Here is my `.eslintrc.json`:
```json
{
  "env": {
    "es2021": true,
    "node": true
  },
  "extends": "eslint:recommended",
  "parserOptions": {
    "ecmaVersion": "latest",
    "sourceType": "module"
  },
  "rules": {
    "indent": [ "error", 2 ],
    "linebreak-style": [ "error", "unix" ],
    "quotes": [ "error", "single" ],
    "semi": [ "error", "always" ],
    "no-empty": ["error", { "allowEmptyCatch": true }]
  }
}
```
The project I have does have quite a bit of code but nothing crazy 
```shell
$ cloc --vcs=git .      
     173 text files.
     172 unique files.                                          
       1 file ignored.

github.com/AlDanial/cloc v 1.92  T=0.14 s (1248.7 files/s, 270968.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                     165           2950           2359          15484
JSON                             3              0              0          12661
SVG                              3              0              0           1395
SQL                              1            768            770            937
-------------------------------------------------------------------------------
SUM:                           172           3718           3129          30477
-------------------------------------------------------------------------------
```
Unfortunately i can't share the entire project since it's private.

Has anyone seen something like this before? Are there any other commands/flags i can run/add to get more details on what exactly might be causing this?

This is extremely annoying when using the eslint VSCode extension since it sometimes freezes the entire IDE for a few seconds.

Edit: tried creating a simple hello world script and linting just that. The indent role still takes over half of the total time. I would assume that the indent role would be one of the simpler ones, no?

Edit2: moee context
Host: Arch Linux with the lts kernel
Eslint version: 8.18.0
Node version: v16.14.2
npm version: 8.5.0

Note: also managed to get the same exact result in an brand new project with just eslint and a hello world script.

**Comment by nzakas** [comment, marked as answer]
Indent is a very complex rule, so it can get slow on larger files. In general, you might be happier using something like Prettier to manage indentation.

**Reply by dumbasPL** [reply]
~~is there a "nice" way to conditionally enable a rule? So i can have it active it active only in ci for example?~~ nevermind, found it.

Thanks for the answer.

The weird thing to me is that wasn't able to find anyone else complain about this so i thought it was an isolated problem. For a rule this slow I would assume someone has already noticed it but was unable to find any evidence of that.

**Reply by nzakas** [reply]
Yeah, it’s hard to know why it’s so slow for you without looking at the specific files you’re running on.
